esptool.py

```diff
@@ -3075,6 +3115,92 @@ def make_image(args):
+            size = self.file.tell()
+            self.file.seek(0)
+            return size
+
+    class MultiBin(object):
+        def __init__(self, name, output_folder='.'):
+            self.name = name
+            self.output_folder = output_folder
+            try:
+                os.makedirs(os.path.realpath(self.output_folder))
+            except OSError:
+                pass    # file already exists, overwrite
+            self.output_path = os.path.realpath(os.path.join(self.output_folder, self.name))
+            self.bin_array = []
+
+        def add_bin(self, addr, file):
+            self.bin_array.append(Bin(addr, file))
+
+        def sort_bin(self):
+            swapped = True
+            while swapped:
+                swapped = False
+                for i in range(len(self.bin_array) - 1):
+                    if self.bin_array[i].addr > self.bin_array[i + 1].addr:
+                        self.bin_array[i], self.bin_array[i + 1] = self.bin_array[i + 1], self.bin_array[i]
+                        swapped = True
+
+        def add_bin_to_other_bin(self, previous, binary):
+            with open(self.output_path, "a") as output_file:
+                output_file.write('\xff' * (binary.addr - previous))
+            print("Add %s from 0x%x to 0x%x (0x%x)" % (
+                binary.file_name, binary.addr, binary.addr + binary.size, binary.size))
+            with open(self.output_path, "a") as output_file:
+                output_file.write(binary.file.read())
+            return binary.addr + binary.size
+
+        def create_bin(self):
+            new_start = 0
+            open(self.output_path, "wb").close()
+            for b in self.bin_array:
+                new_start = self.add_bin_to_other_bin(new_start, b)
+
+        def check_if_possible(self):
+            for i in range(1, len(self.bin_array)):
+                if self.bin_array[i].addr <= (self.bin_array[i - 1].addr + self.bin_array[i - 1].size):
+                    print(self.bin_array[i].addr, (self.bin_array[i - 1].addr + self.bin_array[i - 1].size))
+                    raise Exception("Not possible to create this bin, overlapping between %s and %s" % (
+                        self.bin_array[i].file_name, self.bin_array[i - 1].file_name))
+
+    # Check that bins are provided
+    if len(args.addr_filename) == 0:
+        raise FatalError('No segments specified')
+
+    encrypt = bool(args.keyfile is not None)
+    unenc_name = args.output + '.bin'
+    enc_name = args.output + '.enc.bin'
+
+    # Combine bins into unencrypted image
+    mb = MultiBin(unenc_name)
+    for address, argfile in args.addr_filename:
+        mb.add_bin(address, argfile)
+    mb.sort_bin()
+    mb.check_if_possible()
+    mb.create_bin()
+
+    # Encrypt image (if needed)
+    if encrypt:
+        with open(unenc_name, 'rb') as input_file:
+            with open(enc_name, 'wb') as output_file:
+                espsecure._flash_encryption_operation(
```

Comment:
the underscore prefix in python is for private methods. overall, as this command will probably be run as part of a script or a custom build process then I'd prefer to make encryption a separate step in the script - otherwise when we add ESP32-S2 flash encryption support (for example) this command will also need updating to support that.

Reply:
You are probably right! I am going to leave it as it is for now. But maybe you can change it to serve esptool needs better